ui-kit/src/main/kotlin/org/p2p/uikit/utils/ViewExtensions.kt

```diff
@@ -0,0 +1,168 @@
+package org.p2p.wallet.utils
+
+import android.app.Activity
+import android.content.Context
+import android.content.ContextWrapper
+import android.graphics.Bitmap
+import android.graphics.Canvas
+import android.view.View
+import android.view.ViewGroup
+import android.view.ViewTreeObserver
+import android.view.inputmethod.InputMethodManager
+import android.widget.FrameLayout
+import android.widget.Toast
+import androidx.annotation.ColorRes
+import androidx.annotation.StringRes
+import androidx.appcompat.app.AppCompatActivity
+import androidx.coordinatorlayout.widget.CoordinatorLayout
+import androidx.core.view.WindowInsetsCompat
+import androidx.core.view.doOnAttach
+import androidx.core.view.doOnDetach
+import androidx.fragment.app.Fragment
+import androidx.recyclerview.widget.RecyclerView
+
+fun View.requireActivity(): AppCompatActivity {
+    var context: Context = context
+    while (context is ContextWrapper) {
+        if (context is AppCompatActivity) {
+            return context
+        }
+        context = context.baseContext
+    }
+
+    throw IllegalStateException("View is not attached to any activity")
+}
+
+fun View.focusAndShowKeyboard() {
+    /**
+     * This has to be called when the window already has focus.
+     */
+    fun View.showTheKeyboardNow() {
+        if (isFocused) {
+            post {
+                // We still post the call, just in case we are being notified of the windows focus
+                // but InputMethodManager didn't get properly setup yet.
+                val imm = context.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
+                imm.showSoftInput(this, InputMethodManager.SHOW_IMPLICIT)
+            }
+        }
+    }
+
+    requestFocus()
+
+    if (hasWindowFocus()) {
+        // No need to wait for the window to get focus.
+        showTheKeyboardNow()
+    } else {
+        val listener = ViewTreeObserver.OnWindowFocusChangeListener { hasFocus ->
+            // This notification will arrive just before the InputMethodManager gets set up.
+            if (hasFocus) showTheKeyboardNow()
+        }
+        // We need to wait until the window gets focus.
+        viewTreeObserver.addOnWindowFocusChangeListener(listener)
+
+        doOnDetach { viewTreeObserver.removeOnWindowFocusChangeListener(listener) }
+    }
+}
+
+fun View.createBitmap(): Bitmap {
+    val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
+    val canvas = Canvas(bitmap)
+    draw(canvas)
+    return bitmap
+}
+
+fun Activity.hideKeyboard() {
+    currentFocus?.hideKeyboard()
+}
+
+val View.keyboardIsVisible: Boolean
+    get() = WindowInsetsCompat
+        .toWindowInsetsCompat(rootWindowInsets)
+        .isVisible(WindowInsetsCompat.Type.ime())
+
+fun View.hideKeyboard() {
+    post {
+        (context.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager)
```

Comment:
can be rewritten using ext method: Context:getSystemService<T> that casts automatically